### PR TITLE
ddl: dynamic adjust add index worker number.

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
Currently, if tidb is doing a add index ddl job, then the client change the `tidb_ddl_reorg_worker_cnt` want to speed up add index, but this change will take effect next add index, won't affect current add index job.

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8295)

### CODE REVIEW

**Comment by crazycs520** [comment]
@winkyao @zimulala @ciscoxll PTAL

File: ddl/db_test.go [hunk 1]
```diff
@@ -620,6 +635,21 @@ LOOP:
...
+				c.Assert(err, IsNil)
+				for _, idx := range tbl.Meta().Indices {
+					if idx.Name.L == "c3_index" && idx.State == model.StateWriteReorganization {
+						startReorganization = true
+					}
+				}
+			}
+			if startReorganization {
+				workerCnt := rand.Intn(8) + 8
+				s.mustExec(c, fmt.Sprintf("set @@tidb_ddl_reorg_worker_cnt=%d", workerCnt))
```
Comment (winkyao) [inline ddl/db_test.go]: How can we ensure the worker count is changed to `workerCnt`?

File: ddl/index.go [hunk 2]
```diff
@@ -1053,18 +1053,46 @@ func (w *worker) sendRangeTaskToWorkers(t table.Table, workers []*addIndexWorker
...
+	sessCtx := newContext(reorgInfo.d.store)
+	decodeColMap, err := makeupDecodeColMap(sessCtx, t, indexInfo)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// variable.ddlReorgWorkerCounter can be modified by system variable "tidb_ddl_reorg_worker_cnt".
+	workerCnt := variable.GetDDLReorgWorkerCounter()
+	idxWorkers := make([]*addIndexWorker, 0, workerCnt)
+	defer func() {
```
Comment (winkyao) [inline ddl/index.go]: why not `defer closeAddIndexWorkers(idxWorkers)`?

File: ddl/index.go [hunk 3]
```diff
@@ -1053,18 +1053,46 @@ func (w *worker) sendRangeTaskToWorkers(t table.Table, workers []*addIndexWorker
...
+		// For dynamic adjust add index worker number.
+		workerCnt = variable.GetDDLReorgWorkerCounter()
+		for i := len(idxWorkers); i < int(workerCnt); i++ {
+			sessCtx := newContext(reorgInfo.d.store)
+			idxWorker := newAddIndexWorker(sessCtx, w, i, t, indexInfo, decodeColMap)
+			idxWorker.priority = job.Priority
+			idxWorkers = append(idxWorkers, idxWorker)
+			go idxWorkers[i].run(reorgInfo.d)
+		}
+		if len(idxWorkers) > int(workerCnt) {
```
Comment (winkyao) [inline ddl/index.go]: add some comments here, to describe we are shrinking the worker size.

File: ddl/index.go (hunk 2 above)
Comment (crazycs520) [reply ddl/index.go]:
```go
func main() {
        workers := make([]int, 2)
        defer fmt.Printf("in defer: %v\n", workers)
        workers = append(workers, 1)
        fmt.Printf("in foo  : %v\n", workers)
}
```
the out put is:
```
in foo  : [0 0 1]
in defer: [0 0]
```
So, use `defer closeAddIndexWorkers(idxWorkers)` may can't close some new worker.

File: ddl/db_test.go (hunk 1 above)
Comment (crazycs520) [reply ddl/db_test.go]: em... I check the log when do manual test to know the worker count is changed... 😂

File: ddl/index.go [hunk 4]
```diff
@@ -1037,18 +1037,56 @@ func (w *worker) sendRangeTaskToWorkers(t table.Table, workers []*addIndexWorker
...
+			workers := idxWorkers[workerCnt:]
+			idxWorkers = idxWorkers[:workerCnt]
+			closeAddIndexWorkers(workers)
+		}
+
+		reorgInfo.d.mu.Lock()
+		reorgInfo.d.mu.hook.OnIndexWorkerReorgBefore(len(idxWorkers), len(kvRanges))
+		reorgInfo.d.mu.Unlock()
+
 		log.Infof("[ddl-reorg] start to reorg index of %v region ranges, handle range:[%v, %v).", len(kvRanges), startHandle, endHandle)
```
Comment (ciscoxll) [inline ddl/index.go]: Can I output the `worker number` here?

Comment (crazycs520) [reply ddl/index.go]: nice idea.

Comment (crazycs520) [reply ddl/index.go]: done. PTAL.

**Comment by winkyao** [comment]
@zimulala PTAL

File: ddl/db_test.go [hunk 5]
```diff
@@ -726,7 +728,210 @@ LOOP:
 		delete(handles, h)
 	}
 	c.Assert(handles, HasLen, 0)
+	s.tk.MustExec("drop table test_add_index")
+}
+
+func (s *testDBSuite) TestAddIndexWorkerNum(c *C) {
```
Comment (zimulala) [inline ddl/db_test.go]: Please put it to failtest/fail_db_test.go

File: ddl/testutil/testutil.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
+package testutil
```
Comment (zimulala) [inline ddl/testutil/testutil.go]: Please add the license.

File: ddl/index.go [hunk 7]
```diff
@@ -1100,16 +1081,68 @@ func (w *worker) addPhysicalTableIndex(t table.PhysicalTable, indexInfo *model.I
...
+		// Shrink the worker size.
+		if len(idxWorkers) > int(workerCnt) {
+			workers := idxWorkers[workerCnt:]
+			idxWorkers = idxWorkers[:workerCnt]
+			closeAddIndexWorkers(workers)
+		}
+
+		// gofail: var checkIndexWorkerNum bool
+		//if checkIndexWorkerNum {
+		//	num := int(atomic.LoadInt32(&TestCheckWorkerNumber))
```
Comment (zimulala) [inline ddl/index.go]: Maybe we don't need `atomic.LoadInt3`.

File: ddl/index.go [hunk 8]
```diff
@@ -1100,16 +1081,68 @@ func (w *worker) addPhysicalTableIndex(t table.PhysicalTable, indexInfo *model.I
...
+		}
+		// Shrink the worker size.
+		if len(idxWorkers) > int(workerCnt) {
+			workers := idxWorkers[workerCnt:]
+			idxWorkers = idxWorkers[:workerCnt]
+			closeAddIndexWorkers(workers)
+		}
+
+		// gofail: var checkIndexWorkerNum bool
+		//if checkIndexWorkerNum {
```
Comment (zimulala) [inline ddl/index.go]: s/`//if`/`// if`

**Pull request merged**